### PR TITLE
Introduce type for kinetic energy

### PR DESCRIPTION
**Pull request by xukai92**

### CODE REVIEW

File: src/contrib/ad.jl [hunk 1]
```diff
@@ -1,9 +1,9 @@
 const ADSUPPORT = (:ForwardDiff, :Zygote)
 const ADAVAILABLE = Dict{Module, Function}()
 
-Hamiltonian(metric::M, ℓπ::T, m::Module) where {M<:AbstractMetric,T} = ADAVAILABLE[m](metric, ℓπ)
+Hamiltonian(metric::AbstractMetric, ℓπ::Function, m::Module) = ADAVAILABLE[m](metric, ℓπ)
```
Comment (sethaxen) [inline src/contrib/ad.jl:4]: Do we need this constraint to be a `Function`? This would prevent use of callables.

**Comment by yebai** [comment]
@xukai92 some CI tests are failing - can you take a look?

File: src/contrib/ad.jl (hunk 1 above)
Comment (yebai) [reply src/contrib/ad.jl:4]: It would be good to keep this more generic to support callables.

Comment (xukai92) [reply src/contrib/ad.jl:4]: I wasn’t aware of the callable use cases---will revert it.

**Review by yebai — APPROVED** [review]

**Pull request merged**